### PR TITLE
[feat] Add auto generated multi-issue file

### LINKED ISSUE #12: Auto generate multi-issue file
I'm getting real tired of adding 

#ISSUE#name-task-X#Pass/fail

to a file 15 times so I considered writing a plugin to do that for me but realized I can just add it to the feedback plugin if that's okay with you @slarse 

I'm thinking something like:

`repobee issues feedback create-mi-file`

### PR DESCRIPTION
**Pull request by marceloFA**
Fix #12

Please let me know if any patch to this is needed :)

### CODE REVIEW

File: repobee_feedback/feedback.py [hunk 1]
```diff
@@ -111,6 +112,53 @@ def command(self, api: plug.PlatformAPI):
...
+class GenerateMultiIssueFile(plug.Plugin, plug.cli.Command):
+    __settings__ = plug.cli.command_settings(
+        help="auto generate multi-issue file",
+        description="Will generate a multi-issue file template "
+        "where each pair of student and assignment (args) passed "
+        "will become an issue that starts with the line "
+        "#ISSUE#<STUDENT_REPO_NAME>#<ISSUE_TITLE>, followed by its "
+        "body. Title and body should be filled appropriately later.",
+        category=plug.cli.CoreCommand.issues,
+        action="generate-multi-issue-files",
```
Comment (slarse) [inline repobee_feedback/feedback.py:124]: As we want to test this plugin, we should replace this with an explicit `Action`. It hasn't historically been used explicitly, so it's a bit hidden.

You can define an action like so:

```python
from repobee_plug.cli.categorization import Action

GENERATE_MULTI_ISSUES_FILE_ACTION = Action(
    name="generate-multi-issues-file",
    category=plug.cli.CoreCommand.issues,
)
```

Then you replace both `category` and `action` liike so:

```suggestion
        action=GENERATE_MULTI_ISSUES_FILE_ACTION,
```

We can use this in testing later.

**Review by slarse — CHANGES_REQUESTED** [review]
Hi @marceloFA,

This is a really good first draft! There are quite a few things to fix and touch up, but overall it's very close to the functionality that we've wanted for a long time. Nice job!

So first of all, there's a misconception about what to do with the `assignments` and `students` arguments. In general, in any RepoBee command, the intent of the `assignments` and `students` argument is to operate on each assignment for each student. For example, `--assignments task-1 task-2` and `--students slarse marceloFA` entails operating on `slarse-task-1 slarse-task-2 marceloFA-task-1 marceloFA task-2`. In other words, the cross product of `assignments` and `students`. The function `plug.generate_repo_names(students, assignments)` does that for you (see comments in code).

I would also strongly suggest moving the entire new plugin class and all of its helpers into a separate module (perhaps `_generate_multi_issues_file.py`). Then just add this at the top of the `feedback.py` module (ignoring `F401` as the import is not explicitly used):

```
from repobee_feedback._generate_multi_issues_file import GenerateMultiIssuesFile #  noqa: F401
```

Note that the argument in the `feedback` command is called "multi-issue**s**-file", so make sure that's consistent across the new stuff.

We'll also want to make sure this actually runs (which currently it does not due to the CLI argument collision), so I'd suggest adding at least one test that loads the plugin and asserts that a non-empty output file is created. It could look something like this:

```python
import pathlib
import sys

import pytest

import repobee
import repobee_plug as plug

from repobee_feedback.feedback import (
    GenerateMultiIssueFile,
    GENERATE_MULTI_ISSUES_FILE_ACTION,
    MULTI_ISSUE_FILENAME,
)


def test_creates_non_empty_output_file(tmpdir):
    students = "alice bob".split()
    assignments = "task-1 task-2"
    command = [
        *GENERATE_MULTI_ISSUES_FILE_ACTION.as_name_tuple(),
        "--students",
        students,
        "--assignments",
        assignments,
    ]

    repobee.run(
        command,
        plugins=[GenerateMultiIssueFile],
        workdir=tmpdir,
    )

    outfile = pathlib.Path(tmpdir) / MULTI_ISSUE_FILENAME
    assert outfile.is_file()
    assert len(outfile.read_text(encoding=sys.getdefaultencoding())) > 0
```

Of course, when you move the plugin class to a separate module, you'll need to fix some imports above. But otherwise that test should work.

So that was quite a hefty review, just ask a question if something isn't clear to you!

File: repobee_feedback/feedback.py [hunk 2]
```diff
@@ -111,6 +112,53 @@ def command(self, api: plug.PlatformAPI):
...
+        base_parsers=[plug.BaseParser.ASSIGNMENTS, plug.BaseParser.STUDENTS],
+    )
+
+    assignments = plug.cli.option(required=True)
+    students = plug.cli.option(required=True)
+
+    def command(
+        self,
+    ):
+        _generate_multi_issue_file(self.args.students, self.args.assignments)
```
Comment (slarse) [inline repobee_feedback/feedback.py:134]: I would suggest making this function return a string instead, as that's easier to test.

```suggestion
        content = _generate_multi_issues_file_content(self.args.students, self.args.assignments)
        pathlib.Path(MULTI_ISSUES_FILENAME).write_text(content, encodoing=sys.getdefaultencoding())
```

And then, the only tests you have that actually check that there's an output file are like the ones I suggested for running the entire plugin through RepoBee.

**Comment by slarse** [comment]
I also have a follow-up for this: it would be nice to be able to specify the output file with a CLI argument. But I suggest to finish this first with the current functionality, and then add an `output` CLI option in a separate PR.

**Review by marceloFA — COMMENTED** [review]
Thanks for the detailed review! I learned a lot with these comments :)

**Comment by marceloFA** [comment]
Also, happy holidays for you, @slarse ! <3

File: repobee_feedback/_generate_multi_issues_file.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        "will become an issue that starts with the line "
+        "#ISSUE#<STUDENT_REPO_NAME>#<ISSUE_TITLE>, followed by its "
+        "body. Title and body should be filled appropriately later.",
+        action=GENERATE_MULTI_ISSUES_FILE_ACTION,
+        base_parsers=[plug.BaseParser.ASSIGNMENTS, plug.BaseParser.STUDENTS],
+    )
+
+    def command(
+        self,
+    ):
```
Comment (slarse) [inline repobee_feedback/_generate_multi_issues_file.py:40]: If you add a redundant comma, black will format it like this. So let black add commas instead :)

```suggestion
    def command(self):
```

File: repobee_feedback/_generate_multi_issues_file.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        content = _generate_multi_issues_file_content(
+            self.args.students, self.args.assignments
+        )
+
+        pathlib.Path(MULTI_ISSUES_FILENAME).write_text(
+            "\n\n".join(content),
+            encoding=sys.getdefaultencoding(),
+        )
```
Comment (slarse) [inline repobee_feedback/_generate_multi_issues_file.py:48]: I think it's better to completely generate the content in the function (see comment below), and then just write that. Makes it easier to test it all, and we do want at least one test to verify exactly what the output looks like.

I can see how you arrived at this from my comments, though.

```suggestion
        pathlib.Path(MULTI_ISSUES_FILENAME).write_text(content, encoding=sys.getdefaultencoding())
```

File: repobee_feedback/_generate_multi_issues_file.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+def _generate_multi_issues_file_content(
+    students: List[str], assignments: List[str]
+):
+
+    issue_headers = [
+        f"#ISSUE#{repo_name}#<ISSUE-TITLE>\n<ISSUE-BODY>"
+        for repo_name in plug.generate_repo_names(students, assignments)
+    ]
+
+    return issue_headers
```
Comment (slarse) [inline repobee_feedback/_generate_multi_issues_file.py:62]: Make this return the actual content, and then add at least one test for asserting what the content looks like.

```suggestion
    return "\n\n".join(issue_headers)
```

File: repobee_feedback/_generate_multi_issues_file.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+GENERATE_MULTI_ISSUES_FILE_ACTION = Action(
+    name="generate-multi-issues-file",
+    category=plug.cli.CoreCommand.issues,
+)
+
+
+class GenerateMultiIssuesFile(plug.Plugin, plug.cli.Command):
+    __settings__ = plug.cli.command_settings(
+        help="auto generate multi-issues file",
```
Comment (slarse) [inline repobee_feedback/_generate_multi_issues_file.py:28]: Now that I'm looking at this, it would be good to mention what this file is for.

```suggestion
        help="auto generate multi-issues file for the `issues feedback` command",

```

File: repobee_feedback/_generate_multi_issues_file.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            "\n\n".join(content),
+            encoding=sys.getdefaultencoding(),
+        )
+
+    plug.echo(f"Created multi-issues file '{MULTI_ISSUES_FILENAME}'")
+
+
+def _generate_multi_issues_file_content(
+    students: List[str], assignments: List[str]
+):
```
Comment (slarse) [inline repobee_feedback/_generate_multi_issues_file.py:55]: With the change I suggested to the return value, this should return a string.

```suggestion
) -> str:
```

File: tests/test_generate_multi_issues_file.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    MULTI_ISSUES_FILENAME,
+    GENERATE_MULTI_ISSUES_FILE_ACTION,
+    GenerateMultiIssuesFile,
+)
+
+
+class TestGenerateMultiIssuesFile:
+    """Tests generation of a multi-issues file"""
+
+    def test_creates_non_empty_output_file(self, tmp_path):
```
Comment (slarse) [inline tests/test_generate_multi_issues_file.py:14]: My bad, parts of this test were actually a little bit broken, see comments. It doesn't matter terribly much as the test is just here to ensure that we get any output at all, but it's still nice to have sensible input.

File: tests/test_generate_multi_issues_file.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        students = "alice bob".split()
+        assignments = "task-1 task-2"
```
Comment (slarse) [inline tests/test_generate_multi_issues_file.py:16]:
```suggestion
        assignments = "task-1 task-2".split()
```

File: tests/test_generate_multi_issues_file.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        command = [
+            *GENERATE_MULTI_ISSUES_FILE_ACTION.as_name_tuple(),
+            "--students",
+            students,
```
Comment (slarse) [inline tests/test_generate_multi_issues_file.py:20]:
```suggestion
            *students,
```

File: tests/test_generate_multi_issues_file.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+            "--assignments",
+            assignments,
```
Comment (slarse) [inline tests/test_generate_multi_issues_file.py:22]:
```suggestion
            *assignments,
```

**Review by slarse — CHANGES_REQUESTED** [review]
@marceloFA Very good, just a few touchups and we're home

I suggested some changes. Note that it would also be fine to test the content in the "big test" by just adding an assertion, but it's a little less pure IMO. Writing the content to a file is a different concern than what that content is, so it's nice to be able to test the two independently.

On the other hand, just asserting that the output file contains the correct content encompasses _all_ of the plugin functionality, and so it's a pretty pragmatic test that only differs from the current test by a single assertion. It's also a black box test, which is nice as you can then change the internals any way you want without having to bother with changing the test(s). Honestly, this plugin is so straightforward that I think just replacing the `len` assertion with one that asserts all of the content in the current test is sufficient to test the entire thing, no more granular tests needed. But you decide which way you want to go, as long as there's a test for the actual content I'm satisfied :)

And happy holidays to you too :)

File: repobee_feedback/_generate_multi_issues_file.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        content = _generate_multi_issues_file_content(
+            self.args.students, self.args.assignments
+        )
+
+        pathlib.Path(MULTI_ISSUES_FILENAME).write_text(
+            "\n\n".join(content),
+            encoding=sys.getdefaultencoding(),
+        )
+
+    plug.echo(f"Created multi-issues file '{MULTI_ISSUES_FILENAME}'")
```
Comment (slarse) [inline repobee_feedback/_generate_multi_issues_file.py:50]: Indentation is off, this is part of the class body and not the `command` method.

```suggestion
        plug.echo(f"Created multi-issues file '{MULTI_ISSUES_FILENAME}'")
```

**Review by slarse — CHANGES_REQUESTED** [review]
Missed one thing

**Review by slarse — APPROVED** [review]
Looks good to me, thanks @marceloFA! I'll make a new release of the plugin with these new features later today.
